src/realm.js

```diff
@@ -453,6 +461,99 @@ export class Realm {
     }
   }
 
+  evaluateWithUndoForDiagnostic(f: () => Value): CompilerDiagnostic | Value {
+    if (!this.useAbstractInterpretation) return f();
+    let savedHandler = this.errorHandler;
+    let diagnostic;
+    try {
+      this.errorHandler = d => {
+        diagnostic = d;
+        return "Fail";
+      };
+      let effects = this.evaluateForEffects(f);
+      this.applyEffects(effects);
+      let resultVal = effects[0];
+      if (resultVal instanceof AbruptCompletion) throw resultVal;
+      if (resultVal instanceof PossiblyNormalCompletion) {
+        // in this case one of the branches may complete abruptly, which means that
+        // not all control flow branches join into one flow at this point.
+        // Consequently we have to continue tracking changes until the point where
+        // all the branches come together into one.
+        resultVal = this.composeWithSavedCompletion(resultVal);
+      }
+      invariant(resultVal instanceof Value);
+      return resultVal;
+    } catch (e) {
+      if (diagnostic !== undefined) return diagnostic;
+      throw e;
+    } finally {
+      this.errorHandler = savedHandler;
+    }
+  }
+
+  evaluateForFixpointEffects(
+    loopContinueTest: () => Value,
+    loopBody: () => EvaluationResult
+  ): void | [Effects, Effects] {
+    try {
+      let effects1 = this.evaluateForEffects((loopBody: any));
+      while (true) {
+        this.restoreBindings(effects1[2]);
+        this.restoreProperties(effects1[3]);
+        let effects2 = this.evaluateForEffects(() => {
+          let test = loopContinueTest();
+          if (!(test instanceof AbstractValue)) throw new FatalError("loop terminates before fixed point");
+          return (loopBody(): any);
+        });
+        this.restoreBindings(effects1[2]);
+        this.restoreProperties(effects1[3]);
```

Comment:
Would be easier to read if parts of the tuple was given names with pattern matching.

Reply:
I did not think it worthwhile because the functions strongly suggest what they expect their arguments to be and because Flow types make sure that right tuple column is used.